functions_complete.py

```diff
@@ -0,0 +1,310 @@
+        visible_grid_for_player[guess_y, guess_x] = 2
+        print("Your shot was unsuccesful. This is the computer's grid currently visible to you:")
+        print(visible_grid_for_player)
+        
+        players_turn = False
+        
+    return(players_turn, visible_grid_for_player, computers_ships_uncovered)
+
+def computers_shot(visible_grid_for_computer, players_ships_uncovered):
+    guess_y = random.randrange(0, 10, 1)
+    guess_x = random.randrange(0, 10, 1)
+    
+    if visible_grid_for_computer[guess_y, guess_x] != 0:
+        
+        players_turn = False
+        
+        return(players_turn, visible_grid_for_computer) # computer will start over
+    
+    if (players_grid[guess_y, guess_x] == 1):
+        print("The computer's shot was succesful.")
+        visible_grid_for_computer[guess_y, guess_x] = 1
+        players_ships_uncovered.append(True)
+        
+        players_turn = False # computer has another go
+        
+    elif (players_grid[guess_y, guess_x] != 1):
+        print("The computer's shot was unsuccesful.")
+        visible_grid_for_computer[guess_y, guess_x] = 2
+        
+        players_turn = True # computer missed, so it's the player's turn
+        
+    return(players_turn, visible_grid_for_computer, players_ships_uncovered)
+
+def shoot(computers_grid, players_grid):
+    visible_grid_for_player = np.zeros((10, 10))
+    visible_grid_for_computer = np.zeros((10, 10))
+    players_ships_uncovered = list()
+    computers_ships_uncovered = list()
+
+    # 0 is unknown; 1 is successful shot; 2 is unsuccesful shot; 999 is current selection
+    while (len(players_ships_uncovered) != 18 or len(computers_ships_uncovered) != 18): # until either is a list with 18 items, this loop continues
+    # right now the problem is that the loop doesnt stop once it's sucessful because it first has to break out of the loops below
+        players_turn = True
+        
+        while players_turn == True:
+            print("It is your turn.")
+            players_turn, visible_grid_for_player, computers_ships_uncovered = players_shot(visible_grid_for_player, computers_ships_uncovered)
+            
+        while players_turn == False:
+            print("It is the computer's turn.")
+            players_turn, visible_grid_for_computer, players_ships_uncovered = computers_shot(visible_grid_for_computer, players_ships_uncovered)
+            
+    if (len(players_ships_uncovered) == 18):
+        print("The computer won")
+        print("Your grid:")
+        print(players_grid)
+        print("and the grid visible to the computer:")
+        print(visible_grid_for_computer)
+        
+        return ("Good luck next time.")
+        
+    elif (len(computers_ships_uncovered) == 18):
+        print("You won")
+        print("The computer's grid':")
+        print(computers_grid)
+        print("and what's visible to you':")
+        print(visible_grid_for_player)
+        
+        return ("Congratulations.")
+
```

Comment:
Same over here, I can see that the computer shoots and player shoots are very similar to each other. I think that defining two functions so similar that only discriminate whether it was the computer or the player shooting can be unnecessary. I would create one function for the position of the ship (regardless if it is the computer or the player) and one function for the shoot. Then, you can use the same logic to apply to both scenarios (perhaps the function can take an argument of whether it was the player or the computer shooting; If you ask for input from the player, I think you can save their input in a list or  variable and then use that to scheck against the generated ships and grid)